packages/fork-choice/src/forkChoice/forkChoice.ts

```diff
@@ -161,6 +162,119 @@ export class ForkChoice implements IForkChoice {
     return this.proposerBoostRoot ?? HEX_ZERO_HASH;
   }
 
+  /**
+   * To predict the proposer head of the next slot. That is, to predict if proposer-boost-reorg could happen.
+   * Reason why we can't be certain is because information of the head block is not fully available yet
+   * since the current slot hasn't ended especially the attesters' votes.
+   *
+   * There is a chance we mispredict.
+   *
+   * By calling this function, we assume we are the proposer of next slot
+   *
+   * https://github.com/ethereum/consensus-specs/blob/v1.4.0-beta.4/specs/bellatrix/fork-choice.md#should_override_forkchoice_update
+   */
+  predictProposerHead(headBlock: ProtoBlock, currentSlot?: Slot): ProtoBlock {
+    // Skip re-org attempt if proposer boost (reorg) are disabled
+    if (!this.opts?.proposerBoostEnabled) {
+      this.logger?.verbose("No proposer boot reorg prediction since the related flags are disabled");
+      return headBlock;
+    }
+
+    const parentBlock = this.protoArray.getBlock(headBlock.parentRoot);
+    const proposalSlot = headBlock.slot + 1;
+    currentSlot = currentSlot ?? this.fcStore.currentSlot;
+
+    // No reorg if parentBlock isn't available
+    if (parentBlock === undefined) {
+      return headBlock;
+    }
+
+    const prelimProposerHeadBlock = this.getPreliminaryProposerHead(headBlock, parentBlock, proposalSlot);
+
+    if (prelimProposerHeadBlock === headBlock) {
+      return headBlock;
+    }
+
+    const currentTimeOk = headBlock.slot === currentSlot;
+    if (!currentTimeOk) {
+      return headBlock;
+    }
+
+    this.logger?.info("Current head is weak. Predicting next block to be built on parent of head");
+    return parentBlock;
+  }
+
+  /**
+   *
+   * This function takes in the canonical head block and determine the proposer head (canonical head block or its parent)
+   * https://github.com/ethereum/consensus-specs/pull/3034 for info about proposer boost reorg
+   * This function should only be called during block proposal and only be called after `updateHead()`
+   *
+   * Same as https://github.com/ethereum/consensus-specs/blob/v1.4.0-beta.4/specs/phase0/fork-choice.md#get_proposer_head
+   */
+  getProposerHead(headBlock: ProtoBlock, slot: Slot): ProtoBlock {
```

Comment:
when we consumer this api, would be nice to know the reason head is not reorged even through it's not timely, that will be tracked in a metric at the consumer side

Reply:
> when we consumer this api, would be nice to know the reason head is not reorged even through it's not timely, that will be tracked in a metric at the consumer side

I think not reorging should happen more often than reorg (ie. most of the blocks should arrive on-time than late). So not reorging should be the default case, and reorging should be the special case.

Having a metric tracked for reasons why the default case happens sounds awkward to me. I think `logger.debug` serves this purpose better.